platform/darwin/src/MGLFeature_Private.h

```diff
@@ -37,3 +37,31 @@ mbgl::Feature mbglFeature(mbgl::Feature feature, id identifier, NSDictionary *at
 NS_DICTIONARY_OF(NSString *, id) *NSDictionaryFeatureForGeometry(NSDictionary *geometry, NSDictionary *attributes, id identifier);
 
 NS_ASSUME_NONNULL_END
+
+#define MGLFeatureInitCoding() \
+- (instancetype)initWithCoder:(NSCoder *)decoder { \
+if (self = [super initWithCoder:decoder]) { \
+    NSSet<Class> *identifierClasses = [NSSet setWithArray:@[[NSString class], [NSNumber class]]]; \
+    identifier = [decoder decodeObjectOfClasses:identifierClasses forKey:@"identifier"]; \
+    attributes = [decoder decodeObjectOfClass:[NSDictionary class] forKey:@"attributes"]; \
+} \
+return self; \
+}
+
+#define MGLFeatureEncode() \
+- (void)encodeWithCoder:(NSCoder *)coder { \
+[super encodeWithCoder:coder]; \
+[coder encodeObject:identifier forKey:@"identifier"]; \
+[coder encodeObject:attributes forKey:@"attributes"]; \
+}
+
+#define MGLFeatureIsEqual(klass) \
+- (BOOL)isEqual:(id)other { \
+if (other == self) return YES; \
+if (![other isKindOfClass:[self class]]) return NO; \
+klass *otherFeature = other; \
```

Comment:
`typeof(self)` also works in place of `klass *`, since we’ve already asserted that `other` is of the same kind as `self`.